File: autopayroll-CordaService/workflows-java/src/main/java/net/corda/examples/autopayroll/flows/RequestFlow.java
package net.corda.examples.autopayroll.flows;

import co.paralleluniverse.fibers.Suspendable;
import com.google.common.collect.ImmutableList;
import net.corda.core.contracts.CommandData;
import net.corda.core.flows.*;
import net.corda.core.identity.CordaX500Name;
import net.corda.core.identity.Party;
import net.corda.core.transactions.SignedTransaction;
import net.corda.core.transactions.TransactionBuilder;
import net.corda.core.utilities.ProgressTracker;
import net.corda.core.utilities.ProgressTracker.Step;
import net.corda.examples.autopayroll.contracts.PaymentRequestContract;
import net.corda.examples.autopayroll.states.PaymentRequestState;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

// *********
// * Flows *
// *********
public class RequestFlow {

    @InitiatingFlow
    @StartableByRPC
    public static class RequestFlowInitiator extends FlowLogic<SignedTransaction> {

        private final String amount;
        private final Party towhom;

        private final Step GENERATING_TRANSACTION = new Step("Generating transaction between accounts");
        private final Step PROCESSING_TRANSACTION = new Step("Process transaction with our private key");
        private final Step FINALISING_TRANSACTION = new Step("Obtaining notary signature and recording transaction.");

        // The progress tracker checkpoints each stage of the flow and outputs the specified messages when each
        // checkpoint is reached in the code. See the 'progressTracker.currentStep' expressions within the call()
        // function.
        private final ProgressTracker progressTracker = new ProgressTracker(
                GENERATING_TRANSACTION,
                PROCESSING_TRANSACTION,
                FINALISING_TRANSACTION
        );

        @Nullable
        @Override
        public ProgressTracker getProgressTracker() {
            return progressTracker;
        }

        public RequestFlowInitiator(String amount, Party towhom) {
            this.amount = amount;
            this.towhom = towhom;
        }

        @Suspendable
        @Override
        public SignedTransaction call() throws FlowException {
            progressTracker.setCurrentStep(GENERATING_TRANSACTION);
            Party notary = getServiceHub().getNetworkMapCache().getNotaryIdentities().get(0);
            Party bank = getServiceHub().getNetworkMapCache().getPeerByLegalName(new CordaX500Name("BankOperator", "Toronto", "CA"));
            PaymentRequestState output = new PaymentRequestState(amount, towhom, ImmutableList.of(getOurIdentity(), bank));

            TransactionBuilder txBuilder = new TransactionBuilder(notary);
            CommandData commandData = new PaymentRequestContract.Commands.Request();
            txBuilder.addCommand(commandData, getOurIdentity().getOwningKey(), bank.getOwningKey());
            txBuilder.addOutputState(output, PaymentRequestContract.ID);
            txBuilder.verify(getServiceHub());

            progressTracker.setCurrentStep(PROCESSING_TRANSACTION);
            FlowSession session = initiateFlow(bank);
            SignedTransaction ptx = getServiceHub().signInitialTransaction(txBuilder);
            progressTracker.setCurrentStep(FINALISING_TRANSACTION);
            SignedTransaction stx = subFlow(new CollectSignaturesFlow(ptx, ImmutableList.of(session)));
            return subFlow(new FinalityFlow(stx, ImmutableList.of(session)));
        }
    }

    @InitiatedBy(RequestFlowInitiator.class)
    public static class RequestFlowResponder extends FlowLogic<Void> {

        private final FlowSession counterpartySession;

        public RequestFlowResponder(FlowSession counterPartySession) {
            this.counterpartySession = counterPartySession;
        }

        @Suspendable
        @Override
        public Void call() throws FlowException {
            // Responder flow logic goes here.
            SignedTransaction stx = subFlow(new SignTransactionFlow(counterpartySession) {
                @Override
                protected void checkTransaction(@NotNull SignedTransaction stx) throws FlowException {
                    if (!stx.getInputs().isEmpty()) {
                        throw new FlowException("Payment Request should not have inputs");
                    }
                }
            });

            subFlow(new ReceiveFinalityFlow(counterpartySession, stx.getId()));
            return null;
        }
    }
}

File: autopayroll-CordaService/workflows/src/main/kotlin/net/corda/examples/autopayroll/flows/RequestFlow.kt
package net.corda.examples.autopayroll.flows

import co.paralleluniverse.fibers.Suspendable
import net.corda.examples.autopayroll.contracts.PaymentRequestContract
import net.corda.examples.autopayroll.states.PaymentRequestState
import net.corda.core.contracts.requireThat
import net.corda.core.flows.*
import net.corda.core.utilities.ProgressTracker
import net.corda.core.transactions.TransactionBuilder
import net.corda.core.flows.FlowSession
import net.corda.core.transactions.SignedTransaction
import net.corda.core.flows.FinalityFlow
import net.corda.core.flows.CollectSignaturesFlow
import net.corda.core.identity.CordaX500Name
import net.corda.core.identity.Party


// *********
// * Flows *
// *********
@InitiatingFlow
@StartableByRPC
class RequestFlowInitiator(
        private val amount:String,
        private val towhom:Party
) : FlowLogic<SignedTransaction>() {
    companion object {
        object GENERATING_TRANSACTION : ProgressTracker.Step("Generating transaction for between accounts")
        object PROCESSING_TRANSACTION : ProgressTracker.Step("PROCESS transaction with our private key.")
        object FINALISING_TRANSACTION : ProgressTracker.Step("Obtaining notary signature and recording transaction.")

        fun tracker() = ProgressTracker(
                GENERATING_TRANSACTION,
                PROCESSING_TRANSACTION,
                FINALISING_TRANSACTION
        )
    }

    override val progressTracker = tracker()

    @Suspendable
    override fun call(): SignedTransaction{

        progressTracker.currentStep = GENERATING_TRANSACTION
        val notary = serviceHub.networkMapCache.notaryIdentities[0]
        val bank = serviceHub.networkMapCache.getPeerByLegalName(CordaX500Name("BankOperator", "Toronto", "CA"))!!
        val output = PaymentRequestState(amount, towhom, participants = listOf(ourIdentity, bank))
        val transactionBuilder = TransactionBuilder(notary)
        val commandData = PaymentRequestContract.Commands.Request()
        transactionBuilder.addCommand(commandData, ourIdentity.owningKey, bank.owningKey)
        transactionBuilder.addOutputState(output, PaymentRequestContract.ID)
        transactionBuilder.verify(serviceHub)

        progressTracker.currentStep = PROCESSING_TRANSACTION
        val session = initiateFlow(bank)
        val signedTransaction = serviceHub.signInitialTransaction(transactionBuilder)
        progressTracker.currentStep = FINALISING_TRANSACTION
        val fullySignedTransaction = subFlow(CollectSignaturesFlow(signedTransaction, listOf(session)))
        return subFlow(FinalityFlow(fullySignedTransaction, listOf(session)))

    }
}

@InitiatedBy(RequestFlowInitiator::class)
class RequestFlowResponder(val counterpartySession: FlowSession) : FlowLogic<Unit>() {
    @Suspendable
    override fun call() {
        // Responder flow logic goes here.
        val signTransactionFlow = object : SignTransactionFlow(counterpartySession) {
            override fun checkTransaction(stx: SignedTransaction) = requireThat {
                if (stx.inputs.isNotEmpty()) {
                    throw FlowException("Payment Request should not have inputs.")
                }
            }
        }
        val txId = subFlow(signTransactionFlow).id
        subFlow(ReceiveFinalityFlow(counterpartySession, expectedTxId = txId))

    }
}
